magma/bits.py

```diff
@@ -106,245 +85,522 @@ def __int__(self):
+            return NotImplemented
 
-    def __str__(cls):
-        if cls.isinput():
-            return "In(BFloat[{}])".format(cls.N)
-        if cls.isoutput():
-            return "Out(BFloat[{}])".format(cls.N)
-        return "BFloat[{}]".format(cls.N)
+    def __floordiv__(self, other):
+        try:
+            return self.bvsdiv(other)
+        except InconsistentSizeError as e:
+            raise e from None
+        except TypeError:
+            return NotImplemented
 
-    def qualify(cls, direction):
-        if cls.T.isoriented(direction):
-            return cls
-        return BFloat[cls.N, cls.T.qualify(direction)]
+    def __ge__(self, other):
+        try:
+            return self.bvsge(other)
+        except InconsistentSizeError as e:
+            raise e from None
+        except TypeError:
+            return NotImplemented
 
-    def flip(cls):
-        return BFloat[cls.N, cls.T.flip()]
+    def __gt__(self, other):
+        try:
+            return self.bvsgt(other)
+        except InconsistentSizeError as e:
+            raise e from None
+        except TypeError:
+            return NotImplemented
 
+    def __le__(self, other):
+        try:
+            return self.bvsle(other)
+        except InconsistentSizeError as e:
+            raise e from None
+        except TypeError:
+            return NotImplemented
 
-class BFloat(Bits, metaclass=BFloatKind):
-    def __repr__(self):
-        if not isinstance(self.name, AnonRef):
-            return repr(self.name)
-        ts = [repr(t) for t in self.ts]
-        return 'bfloat([{}])'.format(', '.join(ts))
-
-    def __getitem__(self, key):
-        from .conversions import bfloat
-        result = super().__getitem__(key)
-        if isinstance(key, slice):
-            return bfloat(result)
-        return result
+    def __lt__(self, other):
+        try:
+            return self.bvslt(other)
+        except InconsistentSizeError as e:
+            raise e from None
+        except TypeError:
+            return NotImplemented
+
+    def __rshift__(self, other):
+        try:
+            return self.bvashr(other)
+        except InconsistentSizeError as e:
+            raise e from None
+        except TypeError:
+            return NotImplemented
+
+    def adc(self, other: 'Bits', carry: Bit) -> tp.Tuple['Bits', Bit]:
+        """
+        add with carry
+        returns a two element tuple of the form (result, carry)
+        """
+        T = type(self)
+        other = _coerce(T, other)
+        carry = _coerce(T.unsized_t[1], carry)
+
+        a = self.sext(1)
+        b = other.sext(1)
+        c = carry.zext(T.size)
+
+        res = a + b + c
+        return res[0:-1], res[-1]
+
+    def sext(self, other) -> 'AbstractBitVector':
+        ext = int(other)
+        if ext < 0:
+            raise ValueError()
+
+        T = type(self).unsized_t
+        return self.concat(T[ext]([self[-1] for _ in range(ext)]))
+
+
+class BFloat(Bits):
```

Comment:
Is this still needed?

Reply:
I think it is still used, but IMO should go in some external files which extend the type system.

Reply:
@rsetaluri do you mean that the user code defines this subclass?  Or just to move this to a separate file?

The logic for the BFloat operators is currently handled in the mantle implementations (e.g. the DefineAdd generator dispatches on the type, if it sees a BFloat, it generates a coreir float.add), I think this operator definition will now be moved to this type definition, but could either live in core magma or be moved to user code.

Reply:
at very least it should go to `bfloat.py`...think it's ok to live in core magma then (not sure where else it could go)